File: Python Plotter/CPG.cpp
#include "CPG.hpp"
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

#define MAX(a,b) ((a) > (b) ? (a) : (b))
#define MIN(a,b) ((a) < (b) ? (a) : (b))

void CPG::init(uint8_t nb_modules,
    float frequency,
    float direction,
    float amplc,
    float amplh,
    float nwave,
    float coupling_strength,
    float a_r) {

    //store the number of modules and oscillators
    number_modules = nb_modules;
    number_oscillators = number_modules*2;

    //store the CPG parameters
    param_frequency = frequency;
    param_direction = direction;
    param_amplc = amplc;
    param_amplh = amplh;
    param_nwave = nwave;
    param_coupling_strength = coupling_strength;
    param_a_r = a_r;

    //initialize the coupling and the phase shift matrices
    update_matrices();
    reset();
}

void CPG::step(int8_t* output, float delta_ms) {
    float coupling_term;
    //Update CPG oscillators amplitude and phase
    for(uint8_t i=0;i<number_oscillators;i++) {
        // = compute dtheta = //
        coupling_term = 0;
        for(uint8_t j=0;j<number_oscillators;j++) {
            coupling_term += param_coupling_strength*osc_w[i][j]*osc_r[j]*sin(osc_theta[j]-osc_theta[i]-osc_phi[i][j]);
        }
        osc_dtheta[i] = (2*M_PI*param_frequency + coupling_term);

        // = compute ddr = //
        float ampl;
        float ampl_r;
        //make the amplitude higher for modules further from the head
        //the first module will have "amplc" amplitude and last module will have "amplh"
        if(number_modules > 1) {
            if(i < number_modules) {
                ampl = param_amplh + (param_amplc - param_amplh)/(number_modules-1)*(number_modules-i-1);
            }
            else {
                ampl = param_amplh + (param_amplc - param_amplh)/(number_modules-1)*(2*number_modules-i-1);
            }
        }
        else {
            ampl=param_amplh;
        }
        //change amplitude of left and right oscillators depending on direction
        if(i < number_modules) {
            ampl_r=(ampl-ampl*param_direction)/2.0;
        }
        else {
            ampl_r=(ampl+ampl*param_direction)/2.0;
        }
        osc_ddr[i]  = param_a_r * (0.25*param_a_r * (ampl_r - osc_r[i]) - osc_dr[i]);
    }
    for(uint8_t i=0;i<number_oscillators;i++) {
        //Euler integration
        osc_theta[i] += osc_dtheta[i]*delta_ms/1000.0;
        osc_dr[i] +=  osc_ddr[i]*delta_ms/1000.0;
        osc_r[i] +=   osc_dr[i]*delta_ms/1000.0;
    }
    //Compute joint positions by using left and right oscillators (and convert from radian to degree)
    for(uint8_t i=0;i<number_modules;i++) {
        float setpoint = (osc_r[i+number_modules]*(1.0+cos(osc_theta[i+number_modules])) - osc_r[i]*(1.0+cos(osc_theta[i])))*180/M_PI;
        //set a max angle for each joint
        setpoint = (MAX(setpoint, (-60)));
        output[i] = (int8_t)(MIN(setpoint, (60)));
    }
}

void CPG::reset(void) {
    //reset the oscillators states
    for(uint8_t i=0;i<MAX_OSCILLATORS;i++) {
        osc_r[i] = 0;
        osc_dr[i] = 0;
        osc_ddr[i] = 0;
        osc_theta[i] = 0;
        osc_dtheta[i] = 0;
    }
}

void CPG::set_number_modules(uint8_t nb_modules) {
    number_modules = nb_modules;
    number_oscillators = number_modules*2;
    update_matrices();
    reset();
}

void CPG::set_frequency(float frequency) {
    param_frequency = frequency;
}

void CPG::set_direction(float direction) {
    param_direction = direction;
}

void CPG::set_amplc(float amplc) {
    param_amplc = amplc;
}

void CPG::set_amplh(float amplh) {
    param_amplh = amplh;
}

void CPG::set_nwave(float nwave) {
    param_nwave = nwave;
    update_matrices();
}

void CPG::set_coupling_strength(float coupling_strength) {
    param_coupling_strength = coupling_strength;
}

void CPG::set_a_r(float a_r) {
    param_a_r = a_r;
}

//function to update the phi matrix called when the value of param_nwave or number_modules changes
void CPG::update_matrices(void) {
    //determines the phase shifts between modules to have the desired phase shift between head and end of tail
    float dphi = (param_nwave*2.0*M_PI)/(number_modules);
    //fill the coupling matrix and the phase shift matrix
    for(uint8_t i=0;i<number_oscillators;i++) {
        for(uint8_t j=0;j<number_oscillators;j++) {
            //if the modules of the oscillators are neighbors
            if ((j==(i+1)) && (j!=number_modules)) {
                osc_w[i][j] = 1;
                osc_phi[i][j] = -dphi;
            }
            //if the modules of the oscillators are neighbors
            else if((j==(i-1)) && (j!=number_modules)) {
                osc_w[i][j] = 1;
                osc_phi[i][j] = dphi;
            }
            //if both oscillators are on the same joint
            else if((j==(number_modules+i))) {
                osc_w[i][j] = 1;
                osc_phi[i][j] = M_PI;
            }
            //if both oscillators are on the same joint
            else if((j==(i-number_modules))) {
                osc_w[i][j] = 1;
                osc_phi[i][j] = M_PI;
            }
            else {
                osc_w[i][j] = 0;
                osc_phi[i][j] = 0;
            }
        }
    }
}

File: Python Plotter/CPG.py
"""
 * CPG.py
 * Python implementation of the CPG script from Envirobot V1
 *
 *  Created on: Oct 29, 2024
 *      Author: Séverin Konishi
"""
import numpy as np

#Python implementation of the CPG script from Envirobot V1
class CPG():
    def __init__(self, number_modules, frequency, direction, amplc, amplh, nwave, coupling_strength, a_r):
        # == parameters == #
        self.number_modules = number_modules        #number of modules
        self.number_oscillators = 2*number_modules  #number of oscillators (2 per joint)
        self.frequency = frequency                  #oscillator frequency
        self.direction = direction                  #turning direction (positive/negative) and amplitude
        self.amplc = amplc                          #defines the amplitude of the first joint (closest to head)
        self.amplh = amplh                          #defines the amplitude of the last joint (furthest from head)
        self.nwave = nwave                          #number of waves peaks that can be seen on the robot at the same time
        self.coupling_strength = coupling_strength  #speed at which the phase difference between coupled oscillators converges
        self.a_r = a_r                              #speed at which the amplitude of the oscillators converge

        # == oscillators amplitude and derivatives == #
        self.reset()
        self.update_matrices()

        # == joint angle setpoints == #
        self.output = np.zeros(self.number_modules)

    #compute a discrete step of the CPG controller
    def step(self, delta_ms):
        #Update state of each oscillator
        for i in range(self.number_oscillators):
            coupling = 0
            # = compute dtheta (phase derivative) = #
            for j in range(self.number_oscillators):
                coupling += self.coupling_strength*self.osc_w[i][j]*self.osc_r[j]*np.sin(self.osc_theta[j]-self.osc_theta[i]-self.osc_phi[i][j])
            self.osc_dtheta[i] = (2.0*np.pi*self.frequency + coupling)

            # = compute ddr (amplitude double derivative) = #
            #First module will have an amplitude of "amplc" and last module of "amplh"
            if self.number_modules > 1:
                if i < self.number_modules:
                    ampl = self.amplh + (self.amplc-self.amplh)/(self.number_modules-1)*(self.number_modules-i-1)
                else:
                    ampl = self.amplh + (self.amplc-self.amplh)/(self.number_modules-1)*(2*self.number_modules-i-1)
            else:
                ampl=self.amplh
            #adapt amplitude depending on direction 
            if i < self.number_modules:
                ampl_r=(ampl-ampl*self.direction)/2.0
            else:
                ampl_r=(ampl+ampl*self.direction)/2.0
            self.osc_ddr[i]  = self.a_r * ( 0.25*self.a_r * (ampl_r - self.osc_r[i]) - self.osc_dr[i])

        #Discrete integration
        for i in range(self.number_oscillators):
            self.osc_theta[i] += self.osc_dtheta[i]*(delta_ms/1000.0)
            self.osc_dr[i] += self.osc_ddr[i]*(delta_ms/1000.0)
            self.osc_r[i] += self.osc_dr[i]*(delta_ms/1000.0)

        #Compute joint position
        for i in range(self.number_modules):
            self.output[i] = (self.osc_r[i+self.number_modules]*(1.0+np.cos(self.osc_theta[i+self.number_modules])) - self.osc_r[i]*(1.0+np.cos(self.osc_theta[i])))*180/np.pi
        self.output = np.clip(self.output, a_min=-60, a_max=60)   #limit angle to +- 60 degrees
        return self.output
    
    def reset(self):
        self.osc_r = np.zeros(self.number_oscillators)
        self.osc_dr = np.zeros(self.number_oscillators)
        self.osc_ddr = np.zeros(self.number_oscillators)
        self.osc_theta = np.zeros(self.number_oscillators)
        self.osc_dtheta = np.zeros(self.number_oscillators)

    def update_matrices(self):
        # == coupling weights and phase shift == #
        self.osc_w = np.zeros((self.number_oscillators,self.number_oscillators))    #coupling matrix
        self.osc_phi = np.zeros((self.number_oscillators,self.number_oscillators))  #phase shift matrix
        dphi = self.nwave*2*np.pi/(self.number_modules) #phase shift between oscillators of neighbor modules (to have the desired number of simultaneous wave "peaks" on the robot)
        for i in range(self.number_oscillators):
            for j in range(self.number_oscillators):
                #if the oscillators are from neighbor modules
                if ((j==(i+1)) and (j!=self.number_modules)):
                    self.osc_w[i][j] = 1
                    self.osc_phi[i][j] = -dphi
                #if the oscillators are from neighbor modules
                elif ((j==(i-1)) and (j!=self.number_modules)):
                    self.osc_w[i][j] = 1
                    self.osc_phi[i][j] = dphi
                #if the oscillators are on the same joint
                elif ((j==(self.number_modules+i))):
                    self.osc_w[i][j] = 1
                    self.osc_phi[i][j] = np.pi
                #if the oscillators are on the same joint
                elif ((j==(i-self.number_modules))):
                    self.osc_w[i][j] = 1
                    self.osc_phi[i][j] = np.pi

    def set_number_modules(self, value):
        self.number_modules = value
        self.number_oscillators = value*2
        self.update_matrices()
        self.reset()

    def set_frequency(self, value):
        self.frequency = value

    def set_direction(self, value):
        self.direction = value

    def set_amplc(self, value):
        self.amplc = value

    def set_amplh(self, value):
        self.amplh = value

    def set_nwave(self, value):
        self.nwave = value
        self.update_matrices()

    def set_coupling_strength(self, value):
        self.coupling_strength = value

    def set_a_r(self, value):
        self.a_r = value